Instructor: view results: fix IndexOutOfBoundsException #8415

Haozhe321 opened the pull request:
Fixes #8415 

**Outline of Solution**

- Changed the offending lines. Now the `msqChoices` list will be used to determine the logic for the loop so there will not be a discrepancy between the number of choices and the choices themselves.
- Modified tests to replicate the error before the fix, which passed after the fix.

@craaaa 
<!-- Tell us how you solved the issue. -->
<!-- If there are things you want the reviewers to focus on, include them here as well. -->
<!-- This portion can be skipped if the fix is trivial. -->

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -479,9 +479,9 @@ public String getQuestionAdditionalInfoHtml(int questionNumber, String additiona
+        if (!msqChoices.isEmpty()) {
             optionListHtml.append("<ul style=\"list-style-type: disc;margin-left: 20px;\" >");
-            for (int i = 0; i < numOfMsqChoices; i++) {
+            for (int i = 0; i < msqChoices.size(); i++) {
```
```java
for (String msqChoice : msqChoices.size() {
```

whipermr5 commented:
If we're not removing `numOfMsqChoices`, then this needs to be fixed:
https://github.com/TEAMMATES/teammates/blob/cd9a5365cfed580bae5d1bf15f2994a921e31f03/src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java#L120-L128

Here, `numOfMsqChoices` is set to its generated, non-zero value solely for the purpose of min/max options validation in `validateQuestionDetails` (I think). However, the non-zero value should not be persisted to the datastore, as this makes it inconsistent with `msqChoices`, which is an empty list when persisted to the datastore. Note: both `msqChoices` and `numOfMsqChoices` are stored as a JSON string in the `questionText` field of the `FeedbackQuestion` entity (the field is named `questionMetaData` in `FeedbackQuestionDetails`).

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -479,9 +479,9 @@ public String getQuestionAdditionalInfoHtml(int questionNumber, String additiona
         }
 
-        if (numOfMsqChoices > 0) {
+        if (!msqChoices.isEmpty()) {
```
Actually I feel that for clarity, this whole block should be an else-if of the if statement just above. It should only run for non-generated options.

Haozhe321 commented:
2 possible ways to go from here:
1. Remove `numOfMsqChoices`. This should be fine for future additions as we can rely on `msqChoices.size()`. The only exception is the method `validateQuestionDetails()` that you mentioned, since now we have no ways to check for the violation condition below for msq with generated options:

https://github.com/TEAMMATES/teammates/blob/cd9a5365cfed580bae5d1bf15f2994a921e31f03/src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java#L598-L599

For data migration, I'm not exactly sure of the consequences of that, or how much work it could take. @whipermr5 do you have any recommendations on this?

2. Do not remove `numOfMsqChoices`. In this case we have to make sure that `numOfMsqChoices` and `msqChoices` are consistent in the database. This isn't that big a problem before #8445 and #8689 since we can just store studentlist/teamList/instructorList as the choices; however, after these 2 commits, students will have different choices for their answers so there is no easy way to decide what should be stored in `mcqChoices`. 
What we can do in this case is to store the entire studentList/teamList/instructorList in `mcqChoices`, and set `numOfMsqChoices` to the size of the list. This should be fine since we use `generateOptionsFor` to differentiate between a normal msq and a msq with generated options. So for methods like `validateQuestionDetails()`, we can check for `generateOptionsFor` first and -1 for questions with `generateOptionsFor` when checking for `maxSelectableChoices` violation. For `getQuestionAdditionalInfoHtml()`, we can also check `generateOptionsFor` to determine what goes inside the  additional info box. 

I think `numOfMsqChoices` can remain without causing more problems.

Haozhe321 commented:
Also, just curious if there is any tangible negative impact of the inconsistency in the database? Because if we manage it well, it seems fine (or maybe not haha but I'm interested to know the reason).

Haozhe321 commented:
I tried the approach of storing the entire studentList/teamList/instructorList in `mcqChoices`, and the result is not ideal. Here's an example:

I created a msq with generated options and saved the question. But as you can see from the ss below, the choices will be populated with the entire class list, and that's not what we want. I guess this is why the `mcqChoices` list was set to empty and this answer Bao's [question](https://github.com/TEAMMATES/teammates/issues/8415#issuecomment-364380792).
![image](https://user-images.githubusercontent.com/26651556/37975284-8bb4caae-3211-11e8-8c23-d7dca18952fd.png)

Since this way doesn't seem to work, and we need a field to store the number of msqChoices for validateQuestionDetails(), we can create a new field to store the number of msq choices solely for generated options. @whipermr5 what do you think?

whipermr5 commented:
I think we can keep `numOfMsqChoices` for that purpose, but make it a `transient` field so it doesn't get stored in JSON.

Haozhe321 commented:
@whipermr5 But if we make `numOfMsqChoices` a `transient` field, doesn't that mean we can't use it for its purpose since every time a question is deserialized from the database, `numOfMsqChoices` is set to 0?

whipermr5 commented:
@Haozhe321 I believe validation is only supposed to be done _before_ a question is persisted to the datastore? Anyway auto-generated options are never supposed to be persisted; they should be generated when accessed as the options can change (e.g. if the student list changes).

Haozhe321 commented:
Validation is also done when editing a question i.e. changing the maximum/minimum number of choices, which requires retrieving information of the question from the datastore. 

> Anyway auto-generated options are never supposed to be persisted; they should be generated when accessed as the options can change (e.g. if the student list changes).

I see, that makes sense. Then another way to validate is to use `generateNumOfChoicesForStudentsExcludingSelf()` when validating, instead of using msqChoices. I guess we don't have to use `numOfMsqChoices` then. I'll make some changes(i.e. setting `numOfMsqChoices` to `transient`) and see how that works.

Haozhe321 commented:
@whipermr5 here are the changes made:

1. made `numOfMsqChoices` a `transient` field 
2. changed the parameter for `validateQuestionDetails()` so that it can check the courseId to generate  the size of the generated options. 
3. replaced instances of `numOfMsqChoices` with `msqChoices.size()`
4. Removed methods `generateNumOfChoicesForStudentsExcludingSelf()` and `generateNumOfChoicesForTeamsExcludingSelf()`, replacing them with `numOfChoicesForMsq()` to generated number of choices for every type of option.  


So now the field `numOfMsqChoices` is not used anywhere. That's the first error for PMD. I'm not too sure if removing it is a good idea because questions before this patch would have that field. But as of now, this field is not needed for future questions. 

The 2nd error that PMD flagged out is "Confusing Ternary", which should be skipped when confusing `else if` clauses according to [here](https://sourceforge.net/p/pmd/bugs/1161/) and [here](https://github.com/pmd/pmd/issues/278#issuecomment-285201113).

Haozhe321 commented:
@whipermr5 I'm aware of the lint errors. Need your opinions on them:
1) Should I remove `numOfMsqChoices` completely? this is flagged out in PMD as "Avoid unused private fields".
2) Should I create a new issue to add in the exclusion rule for Confusing Tenary when dealing with `else if`?

Other than the lint errors, the CI tests have passed on my local machine.

whipermr5 commented:
@Haozhe321 Oops didn't read your previous comment carefully 😅 You can always add suppression annotations for now.

wkurniawan07 commented:
@Haozhe321 @whipermr5 I'm currently looking at re-organizing the PMD rules. Consider the request for adding `ignoreElseIf` property in `ConfusingTernary` heard.

Haozhe321 commented:
@whipermr5 fixed the style issues.

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -30,7 +30,7 @@
 
 public class FeedbackMsqQuestionDetails extends FeedbackQuestionDetails {
-    private int numOfMsqChoices;
+    private transient int numOfMsqChoices; //NOPMD
```
Suppress the specific rule and add a comment? E.g.
https://github.com/TEAMMATES/teammates/blob/e32eeff945bc2619d44b3a355e07ac4ad60e0edc/src/main/java/teammates/storage/entity/FeedbackSession.java#L28-L32

Haozhe321 replied:
done!

Haozhe321 commented:
@whipermr5 a reminder to take a look at this :)

whipermr5 commented:
> So now the field numOfMsqChoices is not used anywhere. That's the first error for PMD. I'm not too sure if removing it is a good idea because questions before this patch would have that field. But as of now, this field is not needed for future questions.

Can you investigate what happens if you remove it, then run the new code and view an existing session with questions that have the field? If the deserialisation does not fail, then no harm removing it.

whipermr5 commented:
> I'm able to view past sessions, so it seems fine. Just that questions created before this commit will have numOfMsqChoices as part of their metadata, but it doesn't really affect anything. In the lastest commits I have removed the field in FeedbackMsqQuestionDetails.java.

Sorry for the delay! That's good; I realised Gson gracefully ignores extra unknown fields ([link](http://www.baeldung.com/gson-deserialization-guide)), so we're good to go.

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -341,31 +328,30 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
-
-        return studentList.size() - 1;
-    }
+    private int numOfChoicesForMsq(String courseId, FeedbackParticipantType generatedOption) {
```
`generateOptionsFor` would be a better name.

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -687,7 +674,7 @@ public String validateGiverRecipientVisibility(FeedbackQuestionAttributes feedba
     }
 
     public int getNumOfMsqChoices() {
```
Can we remove this method? The place that calls it can just call `getMsqChoices().size()` instead.

whipermr5 commented on src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java:
```diff
@@ -85,8 +85,9 @@ private void editQuestion(FeedbackQuestionAttributes updatedQuestion) throws Inv
             isError = true;
         }
 
+        String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
```
Use `getNonNullRequestParamValue`

Haozhe321 commented:
@whipermr5 Changes made :) 
There's one place that uses the now-removed `getNumOfMsqChoices()`, I replaced that with the method `getNumOfChoicesForMsq()` so it is more accurate, as getMsqChoices().size() can return 0 when we are using `generatedOptions != ParticipantType.NONE`

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -341,31 +328,30 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
+    public int getNumOfChoicesForMsq(String courseId, FeedbackParticipantType generateOptionsFor) {
+        if (generateOptionsFor == FeedbackParticipantType.NONE) {
+            return msqChoices.size();
+        } else if (generateOptionsFor == FeedbackParticipantType.STUDENTS
```
Can make use of the early returns to avoid else for this entire method.

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -497,6 +482,9 @@ public String getNewQuestionSpecificEditFormHtml() {
     }
 
+    //Confusing Ternary flagged for the `else if` condition below.
+    //Note: Exclusion to this rule will be added in future PMD patch.
```
Space after `//`.

Haozhe321 commented:
@whipermr5 changes made!

whipermr5 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -347,12 +347,10 @@ public int getNumOfChoicesForMsq(String courseId, FeedbackParticipantType genera
             } catch (EntityDoesNotExistException e) {
                 Assumption.fail("Course disappeared");
             }
-        } else {
```
Do it for the whole method - there should be no else ifs.

Haozhe321 replied:
I think I misunderstood what you meant. Do you mean the following?

```java
    public int getNumOfChoicesForMsq(String courseId, FeedbackParticipantType generateOptionsFor) {
        if (generateOptionsFor == FeedbackParticipantType.NONE) {
            return msqChoices.size();
        }

        if (generateOptionsFor == FeedbackParticipantType.STUDENTS
                || generateOptionsFor == FeedbackParticipantType.STUDENTS_EXCLUDING_SELF) {
            List<StudentAttributes> studentList = StudentsLogic.inst().getStudentsForCourse(courseId);
            int sizeOfStudentlist = studentList.size();

            return generateOptionsFor == FeedbackParticipantType.STUDENTS ? sizeOfStudentlist : sizeOfStudentlist - 1;
        }

        if (generateOptionsFor == FeedbackParticipantType.TEAMS
                || generateOptionsFor == FeedbackParticipantType.TEAMS_EXCLUDING_SELF) {
            try {
                List<TeamDetailsBundle> teamList = CoursesLogic.inst().getTeamsForCourse(courseId);
                int sizeOfTeamlist = teamList.size();

                return generateOptionsFor == FeedbackParticipantType.TEAMS ? sizeOfTeamlist : sizeOfTeamlist - 1;
            } catch (EntityDoesNotExistException e) {
                Assumption.fail("Course disappeared");
            }
        }
        List<InstructorAttributes> instructorList = InstructorsLogic.inst().getInstructorsForCourse(courseId);

        return instructorList.size();
    }
```

whipermr5 replied:
Yup!

Haozhe321 replied:
I've push up the changes. What's the difference between this version and the previous version?

whipermr5 replied:
👍 It just looks slightly neater I think
